src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalescingAssignmentTests.cs

```diff
@@ -0,0 +1,1347 @@
+}
+");
+        }
+
+        [Fact]
+        public void DynamicRHS()
+        {
+            CompileAndVerify(@"
+using System;
+using System.Collections.Generic;
+public class C
+{
+    public static void Main()
+    {
+        int? d = null;
+        Console.WriteLine(d ??= GetDynamicClass().Property1);
+        Console.WriteLine(d ??= GetDynamicClass().Property2);
+    }
+
+    public static dynamic GetDynamicClass() => new DynamicClass();
+}
+public class DynamicClass
+{
+    public int Property1
+    {
+        get
+        {
+            Console.WriteLine(""In get_Property1"");
+            return 1;
+        }
+    }
+    public int Property2
+    {
+        get
+        {
+            Console.WriteLine(""In get_Property2"");
+            return 2;
+        }
+    }
+}
+", new[] { CSharpRef }, expectedOutput: @"
+In get_Property1
+1
+1
+");
+        }
+
+        [Fact]
+        public void UseBeforeAssignment()
+        {
+            var compilation = CreateCompilation(@"
+public class C
+{
+    public static void Main()
+    {
+        C c;
+        c ??= new C();
+
+        object x;
+        object y;
+        GetC(x = 1).F ??= (y = 2);
+        x.ToString();
+        y.ToString();
+    }
+    static C GetC(object i) => null;
+    object F { get; set; }
+}").VerifyDiagnostics(
+                // (7,9): error CS0165: Use of unassigned local variable 'c'
+                //         c ??= new C();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "c").WithArguments("c").WithLocation(7, 9),
+                // (13,9): error CS0165: Use of unassigned local variable 'y'
+                //         y.ToString();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(13, 9)
+            );
+
+            var tree = compilation.SyntaxTrees.Single();
+            var semanticModel = compilation.GetSemanticModel(tree);
+            var conditionalAssignment = tree.GetRoot().DescendantNodes()
+                                                      .OfType<AssignmentExpressionSyntax>()
+                                                      .Where(expr => expr.IsKind(SyntaxKind.CoalesceAssignmentExpression))
+                                                      .Skip(1)
+                                                      .Single();
+
+            var yDeclaration = tree.GetRoot().DescendantNodes()
+                                             .OfType<VariableDeclaratorSyntax>()
+                                             .Where(decl => decl.Identifier.Text == "y")
+                                             .Single();
+            var ySymbol = semanticModel.GetDeclaredSymbol(yDeclaration);
+
+            var flowAnalysis = semanticModel.AnalyzeDataFlow(conditionalAssignment);
+            Assert.True(flowAnalysis.Succeeded);
+            Assert.Contains(ySymbol, flowAnalysis.WrittenInside);
+            Assert.DoesNotContain(ySymbol, flowAnalysis.AlwaysAssigned);
+        }
```

Comment:
> Should x.ReadOnlyProperty ??= throw new Exception(); compile successfully?

I don't think so, but I'll add a note to bring it up with LDM

---
In reply to: [199975170](https://github.com/dotnet/roslyn/pull/28253#discussion_r199975170) [](ancestors = 199975170)

Reply:
Actually I don't think it's important to compile that successfully, but it might be interesting to capture the existing behavior in a test.

---
In reply to: [200437600](https://github.com/dotnet/roslyn/pull/28253#discussion_r200437600) [](ancestors = 200437600,199975170)

Reply:
> Consider testing x ??= throw new Exception().

Good catch, this wasn't allowed.

---
In reply to: [200439068](https://github.com/dotnet/roslyn/pull/28253#discussion_r200439068) [](ancestors = 200439068,200437600,199975170)